goas/cl/cl/src/balance.rs

```diff
@@ -1,164 +1,156 @@
+}
+
 impl BalanceWitness {
-    pub fn new(blinding: Scalar) -> Self {
-        Self(blinding)
-    }
+    pub fn random_blinding(mut rng: impl CryptoRngCore) -> [u8; 16] {
+        let mut blinding = [0u8; 16];
+        rng.fill_bytes(&mut blinding);
 
-    pub fn unblinded() -> Self {
-        Self::new(Scalar::ZERO)
+        blinding
     }
 
-    pub fn random(mut rng: impl CryptoRngCore) -> Self {
-        Self::new(Scalar::random(&mut rng))
+    pub fn zero(blinding: [u8; 16]) -> Self {
+        Self {
+            balances: Default::default(),
+            blinding,
+        }
     }
 
-    pub fn commit<'a>(
-        &self,
-        inputs: impl IntoIterator<Item = &'a NoteWitness>,
-        outputs: impl IntoIterator<Item = &'a NoteWitness>,
-    ) -> Balance {
-        let (input_points, input_scalars): (Vec<_>, Vec<_>) = inputs
-            .into_iter()
-            .map(|i| (i.unit, -Scalar::from(i.value)))
-            .unzip();
-
-        let (output_points, output_scalars): (Vec<_>, Vec<_>) = outputs
-            .into_iter()
-            .map(|o| (o.unit, Scalar::from(o.value)))
-            .unzip();
-
-        let points = input_points
-            .into_iter()
-            .chain(output_points)
-            .chain([RISTRETTO_BASEPOINT_POINT]);
-        let scalars = input_scalars
-            .into_iter()
-            .chain(output_scalars)
-            .chain([self.0]);
-
-        let blinded_balance = RistrettoPoint::vartime_multiscalar_mul(scalars, points);
-
-        Balance(blinded_balance)
-    }
-}
+    pub fn from_ptx(ptx: &PartialTxWitness, blinding: [u8; 16]) -> Self {
+        let mut balance = Self::zero(blinding);
 
-pub fn balance(value: u64, unit: Unit, blinding: Scalar) -> Unit {
-    let value_scalar = Scalar::from(value);
-    // can vartime leak the number of cycles through the stark proof?
-    RistrettoPoint::vartime_double_scalar_mul_basepoint(&value_scalar, &unit, &blinding)
-}
+        for input in ptx.inputs.iter() {
+            balance.insert_negative(input.note.unit, input.note.value);
+        }
+
+        for output in ptx.outputs.iter() {
+            balance.insert_positive(output.note.unit, output.note.value);
+        }
+
+        balance.clear_zeros();
+
+        balance
+    }
 
-#[cfg(test)]
-mod test {
-    use super::*;
-    use crate::note::unit_point;
-
-    #[test]
-    fn test_balance_zero_unitless() {
-        // Zero is the same across all units
-        let (nmo, eth) = (unit_point("NMO"), unit_point("ETH"));
-
-        let mut rng = rand::thread_rng();
-        let b = BalanceWitness::random(&mut rng);
-        assert_eq!(
-            b.commit([&NoteWitness::basic(0, nmo)], []),
-            b.commit([&NoteWitness::basic(0, eth)], []),
-        );
+    pub fn insert_positive(&mut self, unit: Unit, value: Value) {
+        let mut found = false;
+        for unit_bal in self.balances.iter_mut() {
+            if unit_bal.unit == unit {
+                found = true;
```

Comment:
can we return instead of break + flag? It's still not constant time

Reply:
yeah, that's better